### PR TITLE
Making LoadResult a union-type.

### PR DESCRIPTION
**Pull request by daiplusplus**
Using a union-type for `LoadResult` to proactively avoid unexpected file load and file-not-found issues. Addresses issue #2340

### CODE REVIEW

File: ILSpy/AssemblyList.cs [hunk 1]
```diff
@@ -121,21 +121,12 @@ public async Task<IList<LoadedAssembly>> GetAllAssemblies()
 
 			foreach (var asm in assemblies)
 			{
-				LoadedAssembly.LoadResult result;
-				try
+				var result = await asm.GetLoadResultAsync();
+				if (result is LoadedAssembly.LoadResult.PackageFallback package)
```
Comment (siegfriedpammer) [inline ILSpy/AssemblyList.cs:125]: `switch` maybe?

File: ILSpy/TreeNodes/AssemblyTreeNode.cs [hunk 2]
```diff
@@ -323,19 +331,19 @@ void HandleException(Exception ex, string message)
 
 			try
 			{
-				var loadResult = LoadedAssembly.GetLoadResultAsync().GetAwaiter().GetResult();
-				if (loadResult.PEFile != null)
+				var loadResult = this.LoadedAssembly.GetLoadResultAsync().GetAwaiter().GetResult();
+				if (loadResult is LoadedAssembly.LoadResult.Successful success)
```
Comment (siegfriedpammer) [inline ILSpy/TreeNodes/AssemblyTreeNode.cs:335]: `switch` maybe?

**Review by siegfriedpammer — COMMENTED** [review]
I think the changes are quite good.

@dgrunwald is there anything you would want to be changed?

@Jehoel do you want to make the changes yourself, or would you be fine with me finishing this PR and merging it?
@Jehoel

**Comment by Jehoel** [comment]
@siegfriedpammer I gather the ILSpy code has changed a lot since I originally submitted this PR - my PR was originally intended as a bugfix-with-benefits rather than a well-written refactoring to enum/union-types.

That said, I'll be happy to resubmit this provided you're okay with me making the _potentially extensive_ changes to adopt enum/union-types throughout ILSpy.

**Comment by dgrunwald** [comment]
I think extensive changes in this direction don't make sense.
C# isn't quite that language yet -- while pattern matching is getting close to make this style of programming usable, the language still lacks [proper union types](https://github.com/dotnet/csharplang/blob/main/proposals/discriminated-unions.md) (and especially Rust-style value union types, which aren't even proposed yet).

**Pull request closed without merging**